src/script-events.c

```diff
@@ -0,0 +1,177 @@
+
+struct bpf_map_def SEC("maps/load_script_map") load_script_map = {
+    .type = BPF_MAP_TYPE_HASH,
+    .key_size = sizeof(u32),
+    .value_size = sizeof(size_t),
+    .max_entries = 1024,
+    .pinning = 0,
+    .namespace = "",
+};
+
+struct bpf_map_def SEC("maps/script_events") script_events = {
+    .type = BPF_MAP_TYPE_PERF_EVENT_ARRAY,
+    .key_size = sizeof(u32),
+    .value_size = sizeof(u32),
+    .max_entries = MAX_TELEMETRY_STACK_ENTRIES * 64,
+    .pinning = 0,
+    .namespace = "",
+};
+
+SEC("kprobe/script_load")
+int kprobe__script_load(struct pt_regs *ctx)
+{
+    struct linux_binprm *bprm = (struct linux_binprm *)PT_REGS_PARM1(ctx);
+    u32 index = (u32)bpf_get_current_pid_tgid();
+
+    bpf_map_update_elem(&load_script_map, &index, &bprm, BPF_ANY);
+
+    return 0;
+}
+
+SEC("kretprobe/ret_script_load")
+int kretprobe__ret_script_load(struct pt_regs *ctx)
+{
+    char br = 0;
+    u32 count = 0;
+    void **bprmp = NULL;
+    void *file_ptr = NULL;
+    u64 loaded = CRC_LOADED;
+    unsigned char *bprm = NULL;
+
+    // Make sure the function succeeded
+    int ret = PT_REGS_RC(ctx);
+    if (ret < 0)
+    {
+        return 0;
+    }
+
+    // Just to be safe 0 out the structs
+    telemetry_event_t ev;
+    __builtin_memset(&ev, 0, sizeof(ev));
+
+    // Initialize some of the telemetry event
+    ev.id = bpf_get_prandom_u32();
+    ev.done = 0;
+    ev.telemetry_type = TE_SCRIPT;
+    ev.u.script_info.mono_ns = bpf_ktime_get_ns();
+
+    // Get Process data and set pid and comm string
+    ev.u.script_info.process.pid = (u32)bpf_get_current_pid_tgid();
+    bpf_get_current_comm(ev.u.script_info.process.comm, sizeof(ev.u.script_info.process.comm));
+
+    // Verify that the offsets are loaded
+    loaded = (u64)bpf_map_lookup_elem(&offsets, &loaded);
+    if (!loaded)
+    {
+        return 0;
+    }
+
+    // Get the stored pointer to the linux_binprm structure and clear it out of the map
+    bprmp = bpf_map_lookup_elem(&load_script_map, &ev.u.script_info.process.pid);
+    if (NULL == bprmp)
+    {
+        bpf_printk("Failed to lookup bprmp\n");
+        return 0;
+    }
+    bpf_map_delete_elem(&load_script_map, &ev.u.script_info.process.pid);
+
+    // Make sure it isn't NULL
+    bprm = (unsigned char *)*bprmp;
+    if (NULL == bprm)
+    {
+        bpf_printk("Failed to deref bprmp\n");
+        return 0;
+    }
+
+    // Get filename
+    ret = read_value(bprm, CRC_BPRM_FILENAME, &file_ptr, sizeof(file_ptr));
+    if (ret < 0)
+    {
+        bpf_printk("Failed to read filename_buf\n");
+        return 0;
+    }
+
+    // Read the filename
+    bpf_probe_read_str(&ev.u.script_info.path, sizeof(ev.u.script_info.path), file_ptr);
+
+    // Output the path
+    bpf_perf_event_output(ctx, &script_events, bpf_get_smp_processor_id(), &ev, sizeof(ev));
+
+    // If the path is / then we don't need to do anything else
+    br = ev.u.script_info.path[0];
+    if (br == '/')
+    {
+        return 0;
+    }
```

Comment:
Is this checking that we got an absolute path right away? Does the shebang accept a relative path to the PWD?

Reply:
So this is to check how the script was run. If you run something like `ls` or `/usr/bin/ls` then it bypasses all that logic below about getting the current working directory because it already has the absolute path. If you run a script like `./scripts/myscript.sh` then it needs to find the current working directory so that it can pass up the absolute path and read the file contents